go/porcelain/deploy_test.go

```diff
@@ -317,10 +317,113 @@ func TestUploadFiles_Errors(t *testing.T) {
+	// require.NoError(t, ioutil.WriteFile(filepath.Join(dir, "foo.html"), []byte("Hello"), 0644))
+
+	// files, err := walk(dir, nil, false, false)
+	// require.NoError(t, err)
+	// d := &models.Deploy{}
+	// for _, bundle := range files.Files {
+	// 	d.Required = append(d.Required, bundle.Sum)
+	// }
+	// // Set SkipRetry to true
+	// err = client.uploadFiles(ctx, d, files, nil, fileUpload, time.Minute, true)
+
+	// // Function upload:
+	hu, _ := url.Parse(server.URL)
+	tr := apiClient.NewWithClient(hu.Host, "/api/v1", []string{"http"}, http.DefaultClient)
+	client := NewRetryable(tr, strfmt.Default, 1)
+	client.uploadLimit = 1
+	apiCtx := context.WithAuthInfo(ctx, apiClient.BearerToken("token"))
+
+	dir, err := ioutil.TempDir("", "deploy")
+	functionsPath := filepath.Join(dir, ".netlify", "functions")
+	os.MkdirAll(functionsPath, os.ModePerm)
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+	require.NoError(t, ioutil.WriteFile(filepath.Join(functionsPath, "foo.js"), []byte("module.exports = () => {}"), 0644))
+
+	files, _, _, err := bundle(ctx, functionsPath, mockObserver{})
+	require.NoError(t, err)
+	d := &models.Deploy{}
+	for _, bundle := range files.Files {
+		d.RequiredFunctions = append(d.RequiredFunctions, bundle.Sum)
+	}
+	// Set SkipRetry to true
+	require.NoError(t, client.uploadFiles(apiCtx, d, files, nil, functionUpload, time.Minute, true))
+	require.Equal(t, err, "[PUT /deploys/{deploy_id}/files/{path}][422] uploadDeployFile default  &{Code:422 Message: Unprocessable Entity}")
+	require.Equal(t, attempts, 1)
+}
+
+func TestUploadFiles400Error_NoSkipRetry(t *testing.T) {
+	attempts := 0
+	ctx := gocontext.Background()
+
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, _ *http.Request) {
+		defer func() {
+			attempts++
+		}()
+
+		rw.WriteHeader(http.StatusUnprocessableEntity)
+		rw.Header().Set("Content-Type", "application/json; charset=utf-8")
+		rw.Write([]byte(`{"message": "Unprocessable Entity", "code": 422 }`))
+		return
+	}))
+	defer server.Close()
+
+	hu, _ := url.Parse(server.URL)
+	tr := apiClient.NewWithClient(hu.Host, "/api/v1", []string{"http"}, http.DefaultClient)
+	client := NewRetryable(tr, strfmt.Default, 1)
+	client.uploadLimit = 1
+	ctx = context.WithAuthInfo(ctx, apiClient.BearerToken("token"))
+
+	// Create some files to deploy
+	dir, err := ioutil.TempDir("", "deploy")
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+	require.NoError(t, ioutil.WriteFile(filepath.Join(dir, "foo.html"), []byte("Hello"), 0644))
+
+	files, err := walk(dir, nil, false, false)
+	require.NoError(t, err)
+	d := &models.Deploy{}
+	for _, bundle := range files.Files {
+		d.Required = append(d.Required, bundle.Sum)
+	}
+	// Set SkipRetry to false
+	err = client.uploadFiles(ctx, d, files, nil, fileUpload, time.Minute, false)
+	// require.Equal(t, err, "[PUT /deploys/{deploy_id}/files/{path}][422] uploadDeployFile default  &{Code:422 Message: Unprocessable Entity}")
+	require.Equal(t, attempts, 12)
```

Comment:
The number of attempts doesn't seem to be consistent, so maybe there's a better way to assert this

Reply:
@4xposed this is one other tiny detail that sometimes makes the test fail - every once in a while it retries 13 times 🤔

Reply:
yes I _think_ (emphasis on _think_) it retries infinitely until the `context` times out or it's cancelled.

I think checking that there's more than 1 attempt (or some arbitrary number that's not too high) might be enough, maybe?

I will try to look more into the retry mechanism tomorrow morning to try to figure out if we can have more control in the test of how many times we retry

Reply:
I added a check to just make sure that the attempts are greater than 1 :) That should probably be sufficient for what we're testing